File: tests/test-resources.js
/**
 * Script de Teste Automatizado - Repinho Compensados (Navegador)
 * Verifica se todos os recursos estão carregando corretamente
 * 
 * Uso:
 * 1. Abra o console do navegador (F12)
 * 2. Cole o conteúdo deste arquivo
 * 3. Execute: runTests()
 */

class ResourceTester {
    constructor() {
        this.results = {
            images: { ok: [], failed: [] },
            stylesheets: { ok: [], failed: [] },
            scripts: { ok: [], failed: [] },
            favicons: { ok: [], failed: [] },
            links: { ok: [], failed: [] }
        };
        this.totalTests = 0;
        this.passedTests = 0;
        this.failedTests = 0;
    }

    // Cores para console
    getColors() {
        return {
            reset: '\x1b[0m',
            bold: '\x1b[1m',
            green: '\x1b[32m',
            red: '\x1b[31m',
            yellow: '\x1b[33m',
            blue: '\x1b[34m',
            cyan: '\x1b[36m'
        };
    }

    // Imprimir cabeçalho
    printHeader(text) {
        const c = this.getColors();
        console.log(`\n${c.bold}${c.cyan}${'='.repeat(70)}${c.reset}`);
        console.log(`${c.bold}${c.cyan}${text.padStart((text.length + 70) / 2)}${c.reset}`);
        console.log(`${c.bold}${c.cyan}${'='.repeat(70)}${c.reset}\n`);
    }

    // Imprimir sucesso
    printSuccess(text) {
        const c = this.getColors();
        console.log(`${c.green}✓${c.reset} ${text}`);
        this.passedTests++;
    }

    // Imprimir erro
    printError(text) {
        const c = this.getColors();
        console.log(`${c.red}✗${c.reset} ${text}`);
        this.failedTests++;
    }

    // Imprimir aviso
    printWarning(text) {
        const c = this.getColors();
        console.log(`${c.yellow}⚠${c.reset} ${text}`);
    }

    // Imprimir info
    printInfo(text) {
        const c = this.getColors();
        console.log(`${c.blue}ℹ${c.reset} ${text}`);
    }

    // Testar se imagem carregou
    testImage(src) {
        return new Promise((resolve) => {
            const img = new Image();
            img.onload = () => resolve(true);
            img.onerror = () => resolve(false);
            img.src = src;
        });
    }

    // Testar se CSS carregou
    testStylesheet(href) {
        return new Promise((resolve) => {
            const link = document.createElement('link');
            link.rel = 'stylesheet';
            link.href = href;
            link.onload = () => resolve(true);
            link.onerror = () => resolve(false);
            document.head.appendChild(link);
        });
    }

    // Testar se script carregou
    testScript(src) {
        return new Promise((resolve) => {
            const script = document.createElement('script');
            script.src = src;
            script.onload = () => resolve(true);
            script.onerror = () => resolve(false);
            document.head.appendChild(script);
        });
    }

    // Testar link HTTP
    testLink(href) {
        return fetch(href, { method: 'HEAD', mode: 'no-cors' })
            .then(() => true)
            .catch(() => false);
    }

    // Extrair recursos da página
    extractResources() {
        const resources = {
            images: [],
            stylesheets: [],
            scripts: [],
            favicons: [],
            links: []
        };

        // Imagens
        document.querySelectorAll('img[src]').forEach(img => {
            resources.images.push({
                src: img.src,
                alt: img.alt || 'Sem alt'
            });
        });

        // Stylesheets
        document.querySelectorAll('link[rel="stylesheet"]').forEach(link => {
            resources.stylesheets.push(link.href);
        });

        // Scripts
        document.querySelectorAll('script[src]').forEach(script => {
            resources.scripts.push(script.src);
        });

        // Favicons
        document.querySelectorAll('link[rel*="icon"]').forEach(link => {
            resources.favicons.push({
                href: link.href,
                sizes: link.sizes || 'N/A'
            });
        });

        // Links internos
        document.querySelectorAll('a[href]').forEach(link => {
            const href = link.href;
            if (href && !href.startsWith('#') && !href.startsWith('javascript:')) {
                resources.links.push({
                    href: href,
                    text: link.textContent.trim().substring(0, 50)
                });
            }
        });

        return resources;
    }

    // Testar imagens
    async testImages(resources) {
        this.printInfo('Testando imagens...');
        
        for (const img of resources.images) {
            this.totalTests++;
            const success = await this.testImage(img.src);
            
            if (success) {
                this.printSuccess(`Imagem: ${img.src}`);
                this.results.images.ok.push(img.src);
            } else {
                this.printError(`Imagem: ${img.src}`);
                this.results.images.failed.push(img.src);
            }
        }
    }

    // Testar stylesheets
    async testStylesheets(resources) {
        this.printInfo('Testando stylesheets...');
        
        for (const href of resources.stylesheets) {
            this.totalTests++;
            try {
                const response = await fetch(href, { method: 'HEAD', mode: 'no-cors' });
                if (response.ok || response.status === 0) {
                    this.printSuccess(`CSS: ${href}`);
                    this.results.stylesheets.ok.push(href);
                } else {
                    this.printError(`CSS: ${href} (Status ${response.status})`);
                    this.results.stylesheets.failed.push(href);
                }
            } catch (e) {
                this.printError(`CSS: ${href} (${e.message})`);
                this.results.stylesheets.failed.push(href);
            }
        }
    }

    // Testar scripts
    async testScripts(resources) {
        this.printInfo('Testando scripts...');
        
        for (const src of resources.scripts) {
            this.totalTests++;
            try {
                const response = await fetch(src, { method: 'HEAD', mode: 'no-cors' });
                if (response.ok || response.status === 0) {
                    this.printSuccess(`Script: ${src}`);
                    this.results.scripts.ok.push(src);
                } else {
                    this.printError(`Script: ${src} (Status ${response.status})`);
                    this.results.scripts.failed.push(src);
                }
            } catch (e) {
                this.printError(`Script: ${src} (${e.message})`);
                this.results.scripts.failed.push(src);
            }
        }
    }

    // Testar favicons
    async testFavicons(resources) {
        this.printInfo('Testando favicons...');
        
        for (const favicon of resources.favicons) {
            this.totalTests++;
            try {
                const response = await fetch(favicon.href, { method: 'HEAD', mode: 'no-cors' });
                if (response.ok || response.status === 0) {
                    this.printSuccess(`Favicon: ${favicon.href} (${favicon.sizes})`);
                    this.results.favicons.ok.push(favicon.href);
                } else {
                    this.printError(`Favicon: ${favicon.href} (Status ${response.status})`);
                    this.results.favicons.failed.push(favicon.href);
                }
            } catch (e) {
                this.printError(`Favicon: ${favicon.href} (${e.message})`);
                this.results.favicons.failed.push(favicon.href);
            }
        }
    }

    // Testar links internos
    async testLinks(resources) {
        this.printInfo('Testando links internos...');
        
        // Limitar a 20 links para não sobrecarregar
        const linksToTest = resources.links.slice(0, 20);
        
        for (const link of linksToTest) {
            this.totalTests++;
            try {
                const response = await fetch(link.href, { method: 'HEAD', mode: 'no-cors' });
                if (response.ok || response.status === 0) {
                    this.printSuccess(`Link: ${link.href}`);
                    this.results.links.ok.push(link.href);
                } else {
                    this.printError(`Link: ${link.href} (Status ${response.status})`);
                    this.results.links.failed.push(link.href);
                }
            } catch (e) {
                this.printError(`Link: ${link.href} (${e.message})`);
                this.results.links.failed.push(link.href);
            }
        }
    }

    // Verificar meta tags
    checkMetaTags() {
        this.printHeader('Verificação de Meta Tags');
        
        const requiredTags = ['viewport', 'description', 'robots'];
        const ogTags = ['og:title', 'og:description', 'og:type', 'og:url'];
        
        this.printInfo('Meta tags essenciais:');
        for (const tag of requiredTags) {
            const meta = document.querySelector(`meta[name="${tag}"]`);
            if (meta) {
                this.printSuccess(`${tag}: ${meta.content.substring(0, 50)}`);
            } else {
                this.printWarning(`${tag}: NÃO ENCONTRADA`);
            }
        }
        
        this.printInfo('Open Graph tags:');
        for (const tag of ogTags) {
            const meta = document.querySelector(`meta[property="${tag}"]`);
            if (meta) {
                this.printSuccess(`${tag}: ${meta.content.substring(0, 50)}`);
            } else {
                this.printWarning(`${tag}: NÃO ENCONTRADA`);
            }
        }
    }

    // Gerar relatório
    generateReport() {
        this.printHeader('Relatório Final de Testes');
        
        console.log(`${this.getColors().bold}Resumo Geral:${this.getColors().reset}`);
        console.log(`  Total de recursos testados: ${this.totalTests}`);
        console.log(`  ${this.getColors().green}✓ Sucesso: ${this.passedTests}${this.getColors().reset}`);
        console.log(`  ${this.getColors().red}✗ Falhas: ${this.failedTests}${this.getColors().reset}`);
        
        if (this.totalTests > 0) {
            const successRate = ((this.passedTests / this.totalTests) * 100).toFixed(1);
            console.log(`\n  ${this.getColors().bold}Taxa de Sucesso: ${successRate}%${this.getColors().reset}`);
        }
        
        console.log(`\n${this.getColors().bold}Detalhes por Tipo:${this.getColors().reset}`);
        
        for (const [type, data] of Object.entries(this.results)) {
            const total = data.ok.length + data.failed.length;
            if (total > 0) {
                const status = data.failed.length === 0 ? this.getColors().green : this.getColors().red;
                console.log(`  ${status}${type.charAt(0).toUpperCase() + type.slice(1)}: ${data.ok.length}/${total} OK${this.getColors().reset}`);
                
                if (data.failed.length > 0) {
                    data.failed.forEach(item => {
                        console.log(`    ${this.getColors().red}✗ ${item}${this.getColors().reset}`);
                    });
                }
            }
        }
        
        if (this.failedTests === 0) {
            console.log(`\n${this.getColors().green}${this.getColors().bold}✓ Todos os testes passaram com sucesso!${this.getColors().reset}`);
        } else {
            console.log(`\n${this.getColors().red}${this.getColors().bold}✗ ${this.failedTests} recursos falharam no teste${this.getColors().reset}`);
        }
    }

    // Executar testes
    async run() {
        this.printHeader('Teste Automatizado de Recursos - Repinho Compensados');
        this.printInfo(`URL: ${window.location.href}`);
        this.printInfo(`Data/Hora: ${new Date().toLocaleString('pt-BR')}`);
        
        const resources = this.extractResources();
        
        await this.testImages(resources);
        await this.testStylesheets(resources);
        await this.testScripts(resources);
        await this.testFavicons(resources);
        await this.testLinks(resources);
        
        this.checkMetaTags();
        this.generateReport();
    }
}

// Função global para executar testes
async function runTests() {
    const tester = new ResourceTester();
    await tester.run();
}

// Executar automaticamente se estiver em desenvolvimento
if (window.location.hostname === 'localhost' || window.location.hostname === '127.0.0.1') {
    console.log('Teste automatizado disponível. Execute: runTests()');
}

File: tests/test-resources.py
#!/usr/bin/env python3
"""
Script de Teste Automatizado - Repinho Compensados
Verifica se todos os links, imagens, favicons, CSS e JS estão carregando corretamente

Uso:
    python3 test-resources.py https://estoque.repinho.ind.br/
    ou
    python3 test-resources.py http://localhost:8000/
"""

import sys
import requests
from urllib.parse import urljoin, urlparse
from bs4 import BeautifulSoup
from collections import defaultdict
import time
from datetime import datetime

# Cores para terminal
class Colors:
    GREEN = '\033[92m'
    RED = '\033[91m'
    YELLOW = '\033[93m'
    BLUE = '\033[94m'
    CYAN = '\033[96m'
    RESET = '\033[0m'
    BOLD = '\033[1m'

def print_header(text):
    """Imprime um cabeçalho formatado"""
    print(f"\n{Colors.BOLD}{Colors.CYAN}{'='*70}{Colors.RESET}")
    print(f"{Colors.BOLD}{Colors.CYAN}{text.center(70)}{Colors.RESET}")
    print(f"{Colors.BOLD}{Colors.CYAN}{'='*70}{Colors.RESET}\n")

def print_success(text):
    """Imprime mensagem de sucesso"""
    print(f"{Colors.GREEN}✓{Colors.RESET} {text}")

def print_error(text):
    """Imprime mensagem de erro"""
    print(f"{Colors.RED}✗{Colors.RESET} {text}")

def print_warning(text):
    """Imprime mensagem de aviso"""
    print(f"{Colors.YELLOW}⚠{Colors.RESET} {text}")

def print_info(text):
    """Imprime mensagem informativa"""
    print(f"{Colors.BLUE}ℹ{Colors.RESET} {text}")

class ResourceTester:
    def __init__(self, base_url):
        self.base_url = base_url.rstrip('/')
        self.session = requests.Session()
        self.session.headers.update({
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
        })
        self.results = defaultdict(list)
        self.stats = {
            'total': 0,
            'success': 0,
            'failed': 0,
            'warnings': 0
        }

    def check_url(self, url, resource_type=""):
        """Verifica se uma URL está acessível"""
        try:
            response = self.session.head(url, timeout=5, allow_redirects=True)
            
            if response.status_code == 200:
                self.stats['success'] += 1
                return True, response.status_code, None
            elif response.status_code in [301, 302, 303, 307, 308]:
                self.stats['warnings'] += 1
                return True, response.status_code, f"Redirecionado para {response.url}"
            else:
                self.stats['failed'] += 1
                return False, response.status_code, f"Status {response.status_code}"
        except requests.exceptions.Timeout:
            self.stats['failed'] += 1
            return False, None, "Timeout na requisição"
        except requests.exceptions.ConnectionError:
            self.stats['failed'] += 1
            return False, None, "Erro de conexão"
        except Exception as e:
            self.stats['failed'] += 1
            return False, None, str(e)

    def extract_resources(self, html_content, page_url):
        """Extrai todos os recursos de uma página HTML"""
        soup = BeautifulSoup(html_content, 'html.parser')
        resources = {
            'links': [],
            'images': [],
            'stylesheets': [],
            'scripts': [],
            'favicons': [],
            'meta': []
        }

        # Links
        for link in soup.find_all('a', href=True):
            href = link['href']
            if href and not href.startswith('#'):
                resources['links'].append((href, link.get_text(strip=True)[:50]))

        # Imagens
        for img in soup.find_all('img', src=True):
            resources['images'].append((img['src'], img.get('alt', 'Sem alt')))

        # Stylesheets
        for link in soup.find_all('link', rel='stylesheet'):
            if link.get('href'):
                resources['stylesheets'].append(link['href'])

        # Scripts
        for script in soup.find_all('script', src=True):
            resources['scripts'].append(script['src'])

        # Favicons
        for link in soup.find_all('link'):
            if link.get('rel') and any('icon' in str(r).lower() for r in link.get('rel', [])):
                if link.get('href'):
                    resources['favicons'].append((link['href'], link.get('sizes', 'N/A')))

        # Meta tags
        for meta in soup.find_all('meta'):
            if meta.get('name') or meta.get('property'):
                name = meta.get('name') or meta.get('property')
                content = meta.get('content', '')
                resources['meta'].append((name, content[:60]))

        return resources

    def test_page(self, page_path):
        """Testa uma página específica"""
        page_url = urljoin(self.base_url, page_path)
        print_info(f"Testando: {page_url}")

        try:
            response = self.session.get(page_url, timeout=10)
            
            if response.status_code != 200:
                print_error(f"Página retornou status {response.status_code}")
                return False

            print_success(f"Página carregada com sucesso")
            resources = self.extract_resources(response.content, page_url)
            
            return resources
        except Exception as e:
            print_error(f"Erro ao acessar página: {str(e)}")
            return False

    def test_resources(self, resources, page_url):
        """Testa todos os recursos extraídos"""
        results = {
            'links': {'ok': 0, 'failed': 0, 'details': []},
            'images': {'ok': 0, 'failed': 0, 'details': []},
            'stylesheets': {'ok': 0, 'failed': 0, 'details': []},
            'scripts': {'ok': 0, 'failed': 0, 'details': []},
            'favicons': {'ok': 0, 'failed': 0, 'details': []},
        }

        # Testar imagens
        print_info("Testando imagens...")
        for img_src, alt_text in resources['images']:
            img_url = urljoin(page_url, img_src)
            success, status, error = self.check_url(img_url, "Imagem")
            self.stats['total'] += 1
            
            if success:
                results['images']['ok'] += 1
                print_success(f"Imagem: {img_src}")
            else:
                results['images']['failed'] += 1
                print_error(f"Imagem: {img_src} - {error}")
                results['images']['details'].append((img_src, error))

        # Testar CSS
        print_info("Testando stylesheets...")
        for css_src in resources['stylesheets']:
            css_url = urljoin(page_url, css_src)
            success, status, error = self.check_url(css_url, "CSS")
            self.stats['total'] += 1
            
            if success:
                results['stylesheets']['ok'] += 1
                print_success(f"CSS: {css_src}")
            else:
                results['stylesheets']['failed'] += 1
                print_error(f"CSS: {css_src} - {error}")
                results['stylesheets']['details'].append((css_src, error))

        # Testar Scripts
        print_info("Testando scripts...")
        for script_src in resources['scripts']:
            script_url = urljoin(page_url, script_src)
            success, status, error = self.check_url(script_url, "Script")
            self.stats['total'] += 1
            
            if success:
                results['scripts']['ok'] += 1
                print_success(f"Script: {script_src}")
            else:
                results['scripts']['failed'] += 1
                print_error(f"Script: {script_src} - {error}")
                results['scripts']['details'].append((script_src, error))

        # Testar Favicons
        print_info("Testando favicons...")
        for favicon_src, sizes in resources['favicons']:
            favicon_url = urljoin(page_url, favicon_src)
            success, status, error = self.check_url(favicon_url, "Favicon")
            self.stats['total'] += 1
            
            if success:
                results['favicons']['ok'] += 1
                print_success(f"Favicon: {favicon_src} ({sizes})")
            else:
                results['favicons']['failed'] += 1
                print_error(f"Favicon: {favicon_src} - {error}")
                results['favicons']['details'].append((favicon_src, error))

        # Testar Links Internos
        print_info("Testando links internos...")
        tested_links = set()
        for link_href, link_text in resources['links']:
            if link_href.startswith('http'):
                continue  # Pular links externos
            
            if link_href in tested_links:
                continue
            tested_links.add(link_href)
            
            link_url = urljoin(page_url, link_href)
            success, status, error = self.check_url(link_url, "Link")
            self.stats['total'] += 1
            
            if success:
                results['links']['ok'] += 1
                print_success(f"Link: {link_href}")
            else:
                results['links']['failed'] += 1
                print_error(f"Link: {link_href} - {error}")
                results['links']['details'].append((link_href, error))

        return results

    def test_meta_tags(self, resources):
        """Verifica meta tags importantes"""
        print_header("Verificação de Meta Tags")
        
        required_tags = ['viewport', 'description', 'robots']
        og_tags = ['og:title', 'og:description', 'og:type', 'og:url']
        
        meta_dict = {name: content for name, content in resources['meta']}
        
        print_info("Meta tags essenciais:")
        for tag in required_tags:
            if tag in meta_dict:
                print_success(f"{tag}: {meta_dict[tag]}")
            else:
                print_warning(f"{tag}: NÃO ENCONTRADA")

        print_info("Open Graph tags:")
        for tag in og_tags:
            if tag in meta_dict:
                print_success(f"{tag}: {meta_dict[tag]}")
            else:
                print_warning(f"{tag}: NÃO ENCONTRADA")

    def generate_report(self, all_results):
        """Gera relatório final"""
        print_header("Relatório Final de Testes")
        
        print(f"{Colors.BOLD}Resumo Geral:{Colors.RESET}")
        print(f"  Total de recursos testados: {self.stats['total']}")
        print(f"  {Colors.GREEN}✓ Sucesso: {self.stats['success']}{Colors.RESET}")
        print(f"  {Colors.RED}✗ Falhas: {self.stats['failed']}{Colors.RESET}")
        print(f"  {Colors.YELLOW}⚠ Avisos: {self.stats['warnings']}{Colors.RESET}")
        
        if self.stats['total'] > 0:
            success_rate = (self.stats['success'] / self.stats['total']) * 100
            print(f"\n  {Colors.BOLD}Taxa de Sucesso: {success_rate:.1f}%{Colors.RESET}")
        
        print(f"\n{Colors.BOLD}Detalhes por Tipo:{Colors.RESET}")
        
        for page_name, results in all_results.items():
            print(f"\n  {Colors.BOLD}{page_name}:{Colors.RESET}")
            for resource_type, data in results.items():
                if resource_type in ['links', 'images', 'stylesheets', 'scripts', 'favicons']:
                    total = data['ok'] + data['failed']
                    if total > 0:
                        status = Colors.GREEN if data['failed'] == 0 else Colors.RED
                        print(f"    {status}{resource_type.capitalize()}: {data['ok']}/{total} OK{Colors.RESET}")
                        
                        if data['details']:
                            for resource, error in data['details']:
                                print(f"      {Colors.RED}✗ {resource}: {error}{Colors.RESET}")

    def run_full_test(self, pages):
        """Executa teste completo em todas as páginas"""
        print_header("Teste Automatizado de Recursos - Repinho Compensados")
        print_info(f"URL Base: {self.base_url}")
        print_info(f"Data/Hora: {datetime.now().strftime('%d/%m/%Y %H:%M:%S')}")
        
        all_results = {}
        
        for page_name, page_path in pages.items():
            print_header(f"Testando: {page_name}")
            
            resources = self.test_page(page_path)
            if not resources:
                continue
            
            results = self.test_resources(resources, urljoin(self.base_url, page_path))
            all_results[page_name] = results
            
            # Testar meta tags apenas na primeira página
            if page_name == list(pages.keys())[0]:
                self.test_meta_tags(resources)
        
        self.generate_report(all_results)
        
        # Retornar status final
        if self.stats['failed'] == 0:
            print_success("\n✓ Todos os testes passaram com sucesso!")
            return True
        else:
            print_error(f"\n✗ {self.stats['failed']} recursos falharam no teste")
            return False

def main():
    if len(sys.argv) < 2:
        print("Uso: python3 test-resources.py <URL_BASE>")
        print("Exemplo: python3 test-resources.py https://estoque.repinho.ind.br/")
        sys.exit(1)

    base_url = sys.argv[1]
    
    # Páginas a testar
    pages = {
        'Página Inicial': 'index.html',
        'Estoque': 'estoque.html',
        'Entenda Produtos': 'entenda-produtos.html',
        'Qualidades': 'nc-qualidades.html',
    }
    
    tester = ResourceTester(base_url)
    success = tester.run_full_test(pages)
    
    sys.exit(0 if success else 1)

if __name__ == '__main__':
    main()
